### PR TITLE
Space-time sampling

### PR DESCRIPTION
**Pull request by dylan-copeland**
Implements space-time sampling and some space-time linear algebra.

### CODE REVIEW

File: Matrix.h [hunk 1]
```diff
@@ -1063,6 +1047,7 @@ class Matrix
     bool d_owns_data;
 };
 
+
```
Comment (kevinhkhuynh) [inline Matrix.h:1050]: extra line

File: STSampling.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#define included_STSAMPLING_h
+
+namespace CAROM {
+
+/**
+ * @brief
+ *
+ * @param[in] s_basis The spatial basis vectors.
+ * @param[in] t_basis The temporal basis vectors.
+ * @param[in] num_samples The number of samples to compute.
```
Comment (kevinhkhuynh) [inline STSampling.h:23]: It seems the doxygen comment needs updating to match with the parameters of the function.

File: STSampling.C [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+{
+    CAROM_VERIFY(t_basis->distributed());
+
+    // Get the number of basis vectors and the size of each basis vector.
+    CAROM_VERIFY(0 < num_f_basis_vectors_used && num_f_basis_vectors_used <= s_basis->numColumns()
+                 && num_f_basis_vectors_used <= t_basis->numColumns());
+    CAROM_VERIFY(num_samples_req > 0);  // TODO: just replace num_samples with num_samples_req
+    const int num_basis_vectors =
+        std::min(num_f_basis_vectors_used, s_basis->numColumns());
+    //const int num_samples = num_samples_req > 0 ? num_samples_req : num_basis_vectors;
```
Comment (kevinhkhuynh) [inline STSampling.C:47]: Do we still need these commented out lines?

File: STSampling.C [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+    CAROM_VERIFY(num_samples <= t_basis->numRows() - numExcluded);
+    const int s_size = s_basis->numRows();
+    const int t_size = t_basis->numRows();
+
+    const int ns_mod_nr = num_samples % num_basis_vectors;
+    int ns = 0;
+
+    // The small matrix inverted by the algorithm.  We'll allocate the largest
+    // matrix we'll need and set its size at each step in the algorithm.
+    Matrix M(num_basis_vectors, num_basis_vectors, false);  // TODO: is this big enough to avoid reallocations?
```
Comment (kevinhkhuynh) [inline STSampling.C:60]: You mentioned below that M should be declared as distributed. Should we do that here, or is this PR only for serial?

**Comment by kevinhkhuynh** [comment]
Will there be any libROM-specific tests for this PR? Or will the tests reside in Laghos?

**Comment by dylan-copeland** [comment]
> Will there be any libROM-specific tests for this PR? Or will the tests reside in Laghos?

For now the testing is just in Laghos. Currently libROM has some small tests for spatial sampling, e.g. DEIM, GNAT. We could add some small tests like those, but it would take a bit of work. I guess the DEIM tests were made just by running DEIM to get output and verifying that output does not change. I don't see a good way to set up a test without using the code to generate results in the first place.

File: STSampling.C (hunk 4 above)
Comment (dylan-copeland) [reply STSampling.C:60]: This PR is only for the serial case. There will be a significant amount of work to get the parallel case, so I am just trying to document some of the things to do.

File: STSampling.C (hunk 3 above)
Comment (dylan-copeland) [reply STSampling.C]: Removed.

File: STSampling.h (hunk 2 above)
Comment (dylan-copeland) [reply STSampling.h]: Updated.

**Comment by chldkdtn** [comment]
> > Will there be any libROM-specific tests for this PR? Or will the tests reside in Laghos?
> 
> For now the testing is just in Laghos. Currently libROM has some small tests for spatial sampling, e.g. DEIM, GNAT. We could add some small tests like those, but it would take a bit of work. I guess the DEIM tests were made just by running DEIM to get output and verifying that output does not change. I don't see a good way to set up a test without using the code to generate results in the first place.

I am okay not to have a unit test in this PR, but in next PR, please add unit test that is similar to the ones for DEIM and GNAT.

**Review by chldkdtn — APPROVED** [review]

**Review by siuwuncheung — APPROVED** [review]

**Review by kevinhkhuynh — APPROVED** [review]

**Pull request merged**